File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/ssh/SshWrappedDestination.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.base.ssh;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.airbyte.commons.json.Jsons;
import io.airbyte.commons.resources.MoreResources;
import io.airbyte.integrations.base.AirbyteMessageConsumer;
import io.airbyte.integrations.base.Destination;
import io.airbyte.protocol.models.AirbyteConnectionStatus;
import io.airbyte.protocol.models.AirbyteMessage;
import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
import io.airbyte.protocol.models.ConnectorSpecification;
import java.util.List;
import java.util.function.Consumer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Decorates a Destination with an SSH Tunnel using the standard configuration that Airbyte uses for
 * configuring SSH.
 */
public class SshWrappedDestination implements Destination {

  private static final Logger LOGGER = LoggerFactory.getLogger(SshWrappedDestination.class);

  private final Destination delegate;
  private final List<String> hostKey;
  private final List<String> portKey;
  private final String endPointKey;

  public SshWrappedDestination(final Destination delegate,
                               final List<String> hostKey,
                               final List<String> portKey) {
    this.delegate = delegate;
    this.hostKey = hostKey;
    this.portKey = portKey;
    this.endPointKey = null;
  }

  public SshWrappedDestination(final Destination delegate,
                               final String endPointKey) {
    this.delegate = delegate;
    this.endPointKey = endPointKey;
    this.portKey = null;
    this.hostKey = null;
  }

  @Override
  public ConnectorSpecification spec() throws Exception {
    // inject the standard ssh configuration into the spec.
    final ConnectorSpecification originalSpec = delegate.spec();
    final ObjectNode propNode = (ObjectNode) originalSpec.getConnectionSpecification().get("properties");
    propNode.set("tunnel_method", Jsons.deserialize(MoreResources.readResource("ssh-tunnel-spec.json")));
    return originalSpec;
  }

  @Override
  public AirbyteConnectionStatus check(final JsonNode config) throws Exception {
    return (endPointKey != null) ? SshTunnel.sshWrap(config, endPointKey, delegate::check)
        : SshTunnel.sshWrap(config, hostKey, portKey, delegate::check);
  }

  @Override
  public AirbyteMessageConsumer getConsumer(final JsonNode config,
                                            final ConfiguredAirbyteCatalog catalog,
                                            final Consumer<AirbyteMessage> outputRecordCollector)
      throws Exception {
    final SshTunnel tunnel = (endPointKey != null) ? SshTunnel.getInstance(config, endPointKey) : SshTunnel.getInstance(config, hostKey, portKey);

    final AirbyteMessageConsumer delegateConsumer;
    try {
      delegateConsumer = delegate.getConsumer(tunnel.getConfigInTunnel(), catalog, outputRecordCollector);
    } catch (final Exception e) {
      LOGGER.error("Exception occurred while getting the delegate consumer, closing SSH tunnel", e);
      tunnel.close();
      throw e;
    }
    return AirbyteMessageConsumer.appendOnClose(delegateConsumer, tunnel::close);
  }

}

File: airbyte-cdk/java/airbyte-cdk/core/bin/main/io/airbyte/cdk/integrations/base/ssh/SshWrappedDestination.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.base.ssh

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ObjectNode
import io.airbyte.cdk.integrations.base.AirbyteMessageConsumer
import io.airbyte.cdk.integrations.base.AirbyteTraceMessageUtility
import io.airbyte.cdk.integrations.base.Destination
import io.airbyte.cdk.integrations.base.SerializedAirbyteMessageConsumer
import io.airbyte.commons.concurrency.VoidCallable
import io.airbyte.commons.json.Jsons
import io.airbyte.commons.resources.MoreResources
import io.airbyte.protocol.models.v0.AirbyteConnectionStatus
import io.airbyte.protocol.models.v0.AirbyteMessage
import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog
import io.airbyte.protocol.models.v0.ConnectorSpecification
import io.github.oshai.kotlinlogging.KotlinLogging
import java.util.*
import java.util.function.Consumer

private val LOGGER = KotlinLogging.logger {}
/**
 * Decorates a Destination with an SSH Tunnel using the standard configuration that Airbyte uses for
 * configuring SSH.
 */
class SshWrappedDestination : Destination {
    private val delegate: Destination
    private val hostKey: List<String>?
    private val portKey: List<String>?
    private val endPointKey: String?

    constructor(delegate: Destination, hostKey: List<String>, portKey: List<String>) {
        this.delegate = delegate
        this.hostKey = hostKey
        this.portKey = portKey
        this.endPointKey = null
    }

    constructor(delegate: Destination, endPointKey: String) {
        this.delegate = delegate
        this.endPointKey = endPointKey
        this.portKey = null
        this.hostKey = null
    }

    @Throws(Exception::class)
    override fun spec(): ConnectorSpecification {
        // inject the standard ssh configuration into the spec.
        val originalSpec = delegate.spec()
        val propNode = originalSpec.connectionSpecification["properties"] as ObjectNode
        propNode.set<JsonNode>(
            "tunnel_method",
            Jsons.deserialize(MoreResources.readResource("ssh-tunnel-spec.json"))
        )
        return originalSpec
    }

    @Throws(Exception::class)
    override fun check(config: JsonNode): AirbyteConnectionStatus? {
        try {
            return if ((endPointKey != null))
                SshTunnel.sshWrap(config, endPointKey) { c: JsonNode -> delegate.check(c) }
            else
                SshTunnel.sshWrap(config, hostKey!!, portKey!!) { c: JsonNode -> delegate.check(c) }
        } catch (e: RuntimeException) {
            val sshErrorMessage =
                "Could not connect with provided SSH configuration. Error: " + e.message
            AirbyteTraceMessageUtility.emitConfigErrorTrace(e, sshErrorMessage)
            return AirbyteConnectionStatus()
                .withStatus(AirbyteConnectionStatus.Status.FAILED)
                .withMessage(sshErrorMessage)
        }
    }

    @Throws(Exception::class)
    override fun getConsumer(
        config: JsonNode,
        catalog: ConfiguredAirbyteCatalog,
        outputRecordCollector: Consumer<AirbyteMessage>
    ): AirbyteMessageConsumer? {
        val tunnel = getTunnelInstance(config)

        val delegateConsumer: AirbyteMessageConsumer?
        try {
            delegateConsumer =
                delegate.getConsumer(tunnel.configInTunnel, catalog, outputRecordCollector)
        } catch (e: Exception) {
            LOGGER.error(e) {
                "Exception occurred while getting the delegate consumer, closing SSH tunnel"
            }
            tunnel.close()
            throw e
        }
        return AirbyteMessageConsumer.Companion.appendOnClose(
            delegateConsumer,
            VoidCallable { tunnel.close() }
        )
    }

    @Throws(Exception::class)
    override fun getSerializedMessageConsumer(
        config: JsonNode,
        catalog: ConfiguredAirbyteCatalog,
        outputRecordCollector: Consumer<AirbyteMessage>
    ): SerializedAirbyteMessageConsumer? {
        val clone = Jsons.clone(config)
        val connectionOptionsConfig: Optional<JsonNode> =
            Jsons.getOptional(clone, SshTunnel.Companion.CONNECTION_OPTIONS_KEY)
        if (connectionOptionsConfig.isEmpty) {
            LOGGER.info { "No SSH connection options found, using defaults" }
            if (clone is ObjectNode) { // Defensive check, it will always be object node
                val connectionOptions = clone.putObject(SshTunnel.Companion.CONNECTION_OPTIONS_KEY)
                connectionOptions.put(
                    SshTunnel.Companion.SESSION_HEARTBEAT_INTERVAL_KEY,
                    SshTunnel.Companion.SESSION_HEARTBEAT_INTERVAL_DEFAULT_IN_MILLIS
                )
                connectionOptions.put(
                    SshTunnel.Companion.GLOBAL_HEARTBEAT_INTERVAL_KEY,
                    SshTunnel.Companion.GLOBAL_HEARTBEAT_INTERVAL_DEFAULT_IN_MILLIS
                )
            }
        }
        val tunnel = getTunnelInstance(clone)
        val delegateConsumer: SerializedAirbyteMessageConsumer?
        try {
            delegateConsumer =
                delegate.getSerializedMessageConsumer(
                    tunnel.configInTunnel,
                    catalog,
                    outputRecordCollector
                )
        } catch (e: Exception) {
            LOGGER.error(e) {
                "Exception occurred while getting the delegate consumer, closing SSH tunnel"
            }
            tunnel.close()
            throw e
        }
        return SerializedAirbyteMessageConsumer.Companion.appendOnClose(
            delegateConsumer,
            VoidCallable { tunnel.close() }
        )
    }

    @Throws(Exception::class)
    protected fun getTunnelInstance(config: JsonNode): SshTunnel {
        return if ((endPointKey != null)) SshTunnel.Companion.getInstance(config, endPointKey)
        else SshTunnel.Companion.getInstance(config, hostKey!!, portKey!!)
    }

    override val isV2Destination: Boolean
        get() = delegate.isV2Destination

    companion object {}
}
